Add sqlite locking blog post (#1399)
jellyfin/jellyfin.org: blog/2025/05-20-sqlite-locking/index.mdx

This seems to be a not uncommon issue there are many reports to be found researching the issue.

The factor that makes this issue so bad is that it does not happen reliably. So far we only have one Team member where this can be (somewhat) reliably be reproduced makes this even worse a bug. From the reports this issue happens across all Operating systems, HDD speeds and with or without virtualization. So we do not have any deciding factor identified that even contributes to the likeliness of the issue happening.

## The solution

Since we moved the codebase over to EFCore proper, we have the tools to actually do something about this as EFCore gives us a structured abstraction level.
EFCore supports a way of hooking into every command execution or transaction creating, called (Interceptors)[https://learn.microsoft.com/en-us/ef/core/logging-events-diagnostics/interceptors].

> With an interceptor we can do finally do the straight forward idea of just "not" writing to the database in parallel in an transparent way for the caller. 

Reviewer: ```suggestion
With an interceptor we can finally do the straight forward idea of just "not" writing to the database in parallel in a transparent way to the caller. 
```

> The overall idea is to have multiple strategies of locking, because all levels of synchronization will inevitable at the cost of performance we only want to do it when its really necessary.

Reviewer: ```suggestion
The overall idea is to have multiple strategies of locking. Because all levels of synchronization will inevitably come at the cost of performance, we only want to do it when it is really necessary.
```

Both the Optimistic and Pessimistic locking behaviors work on the same principle, they both create two Interceptors one for handling transactions and one for handling commands itself and also override the SaveChanges method of the `JellyfinDbContext`.

### Optimistic locking behavior 

Optimistic locking means to assume the operation in question will succeed and only handle issues afterwards. In essence this can be boiled down to "Try and Retry and Retry ..." for a set number of times until either we succeed with the operation or fail entirely. 
This still leaves the possibility that we do not actually be able to perform an write but the introduced overhead is far less then the Pessimistic locking behavior.

The idea why this just works out fine is that every time two operations try to write to the database, one will always just win without issue so the 2nd operation that will fail to write at the first try will just wait for some time and the retry a few times.

Jellyfin uses the `Polly` library perform the retry behavior and will only retry operations it will find have been locked due to this exact issue.

### Pessimistic locking behavior

> Pessimistic locking always locks when a write to SQLite should be performed. Essentially every time an Transaction is started or a write operation on the database is done though EFCore, jellyfin will wait until all other read operations are finished and then block all other operations may they be read or write until the write in question has been performed.

Reviewer: ```suggestion
Pessimistic locking always locks when a write to SQLite should be performed. Essentially every time an Transaction is started or a write operation on the database is done though EFCore, Jellyfin will wait until all other read operations are finished and then block all other operations may they be read or write until the write in question has been performed.
```

> This however means, that jellyfin can only ever perform a single write to the database, even if it would technically does not need to. 

Reviewer: ```suggestion
This however means, that Jellyfin can only ever perform a single write to the database, even if it would technically does not need to. 
```

> In theory, an application should have no issue reading from table "Alice" while writing to table "Bob" however to eliminate all possible sources of concurrency related locking, in this mode jellyfin will only ever allow a single write performed on its database.

Reviewer: ```suggestion
In theory, an application should have no issue reading from table "Alice" while writing to table "Bob" however to eliminate all possible sources of concurrency related locking, Jellyfin will only ever allow a single write performed on its database in this mode.
```

While this will absolutely result in the most stable operation, it will undoubtedly also be the slowest. 

Jellyfin uses a [ReaderWriterLockSlim](https://learn.microsoft.com/en-us/dotnet/fundamentals/runtime-libraries/system-threading-readerwriterlockslim) to lock the operations, that means we allow an unlimited number of reads to happen concurrently while only one write may ever be done on the database.

### The future Smart locking behavior

In the future we might also consider combining both modes, to get the best of both worlds. 

> 

Reviewer: why 2 blank rows

# The result

> Initial testing showed that with both modes, we had great success in handling the underlying issue. While we are not yet sure why this happens only on some systems when others work, we at least now have an option for users previously left out of using jellyfin.

Reviewer: ```suggestion
Initial testing showed that with both modes, we had great success in handling the underlying issue. While we are not yet sure why this happens only on some systems when others work, we at least now have an option for users previously left out of using Jellyfin.
```
